### PR TITLE
add: filter, search and bookmark functionality

### PR DESCRIPTION
**Pull request by ApurvaSawant11**
In this PR, I have added the below features:
- Add/Remove post and questions to Bookmarks
- Filter threads by tags 
- Filter threads by Title or Content
- Sort threads by Most Upvotes
- Sort threads by latest creation date
- Search users by their name

### CODE REVIEW

File: src/features/singleQuestion/SingleQuestion.jsx [hunk 1]
```diff
@@ -163,21 +168,41 @@ const SingleQuestion = () => {
...
             </div>
-            <div className="flex items-center gap-2 cursor-pointer">
-              <BookmarkOutlineIcon size={24} />{" "}
+            <div
+              className="flex items-center gap-2 cursor-pointer"
+              onClick={() =>
+                dispatch(
+                  addOrRemoveQueBookmark({
+                    questionId: question._id,
+                    isBookmark: isBookmarked ? false : true,
```
Comment (foolhardy21) [inline src/features/singleQuestion/SingleQuestion.jsx:196]: appreciation - nice way of including both actions in the same dispatch.

File: src/components/modal/searchModal/SearchModal.jsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  return (
+    <>
+      <div
+        className={`flex items-start justify-center fixed top-0 bottom-0 left-0 right-0 bg-overlay`}
+        onClick={() => setShowSearchModal(false)}
+      >
+        <div
+          onClick={(event) => {
+            event.stopPropagation();
```
Comment (foolhardy21) [inline src/components/modal/searchModal/SearchModal.jsx]: I wanted to know as there is no event handler for this `onClick`, then why have you added `stopPropogation()`?

File: src/components/header/Header.jsx [hunk 3]
```diff
@@ -51,6 +66,23 @@ const Header = () => {
...
             Profile
           </li>
+          <li
+            className="hover:bg-slate-200 flex items-center gap-4  px-3 py-1 rounded-md cursor-pointer"
+            tabIndex="0"
+            onClick={() => {
+              navigate("/bookmarks");
+              setIsShowDropDown(false);
+            }}
+            onKeyDown={(event) => {
```
Comment (foolhardy21) [inline src/components/header/Header.jsx:76]: appreciation - good use of `keydown` event

File: src/components/filterBar/FilterBar.jsx [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  const [filterCategory, setFilterCategory] = useState("none");
+  const [filterInput, setFilterInput] = useState("");
+
+  useEffect(() => {
+    if (filterCategory !== "none") {
+      filterCategory === "tags" ? filterByTags() : filterByContent();
+    } else if (sortCategory === "most-upvotes") {
+      sortByUpvotes();
+    } else {
+      console.log("------here----");
```
Comment (foolhardy21) [inline src/components/filterBar/FilterBar.jsx]: `console.log`

**Review by foolhardy21 — APPROVED** [review]
good practices have been used throughout the code. gave some suggestions, otherwise good to go ahead.

File: src/components/modal/searchModal/SearchModal.jsx (hunk 2 above)
Comment (ApurvaSawant11) [reply src/components/modal/searchModal/SearchModal.jsx]: Right. Had a different modal structure before. Forgot to remove this. Thanks for pointing it out

File: src/components/filterBar/FilterBar.jsx (hunk 4 above)
Comment (ApurvaSawant11) [reply src/components/filterBar/FilterBar.jsx]: Yeah. Will remove this

File: src/App.css [hunk 5]
```diff
@@ -14,4 +14,52 @@
     background: linear-gradient(180deg, #08AEEA 0%, #2AF598 100%);
     -webkit-background-clip: text;
     -webkit-text-fill-color: transparent;
-  }
\ No newline at end of file
+}
+
+.styles_chip-input__3aumB .styles_chip__i-PT7 {
```
Comment (vrx29) [inline src/App.css:19]: why are your classnames like __3aumb. Are these system/library generated or you wrote these on your own?. If you wrote them then please change them as these are not easily understandable

**Review by vrx29 — APPROVED** [review]
Everything looks perfect. Nice folder structure. You can merge it.
Just one thing your search bar is not functioning. It gets close when we click on the search input. Please if this issue is on my end or for everyone else

File: src/App.css (hunk 5 above)
Comment (ApurvaSawant11) [reply src/App.css:19]: These are library-generated names. I have used them to overwrite default styling

**Comment by ApurvaSawant11** [comment]
> Everything looks perfect. Nice folder structure. You can merge it. Just one thing your search bar is not functioning. It gets close when we click on the search input. Please if this issue is on my end or for everyone else

I have fixed the issue.  Thanks for reviewing.

**Pull request merged**